pkg/controller/cloudfront/distribution/setup.go

```diff
@@ -0,0 +1,182 @@
+				kube: mgr.GetClient(),
+				opts: []option{
+					func(e *external) {
+						e.preCreate = preCreate
+						e.postCreate = postCreate
+						e.lateInitialize = lateInitialize
+						e.preObserve = preObserve
+						e.postObserve = postObserve
+						e.isUpToDate = isUpToDate
+						e.preUpdate = preUpdate
+						e.preDelete = preDelete
+					},
+				},
+			}),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+func preCreate(_ context.Context, cr *svcapitypes.Distribution, cdi *svcsdk.CreateDistributionInput) error {
+	cdi.DistributionConfig.CallerReference = awsclients.String(string(cr.UID))
+	cdi.DistributionConfig.Origins.Quantity =
+		awsclients.PInt64Value(len(cr.Spec.ForProvider.DistributionConfig.Origins.Items))
+
+	return nil
+}
+
+func postCreate(_ context.Context, cr *svcapitypes.Distribution, cdo *svcsdk.CreateDistributionOutput,
+	ec managed.ExternalCreation, err error) (managed.ExternalCreation, error) {
+	if err != nil {
+		return managed.ExternalCreation{}, err
+	}
+
+	meta.SetExternalName(cr, awsclients.StringValue(cdo.Distribution.Id))
+	ec.ExternalNameAssigned = true
+
+	return ec, nil
+}
+
+func preObserve(_ context.Context, cr *svcapitypes.Distribution, gdi *svcsdk.GetDistributionInput) error {
+	gdi.Id = awsclients.String(meta.GetExternalName(cr))
+
+	return nil
+}
+
+func lateInitialize(in *svcapitypes.DistributionParameters, gdo *svcsdk.GetDistributionOutput) error {
+	inConfig, respConfig := in.DistributionConfig, gdo.Distribution.DistributionConfig
+
+	_, err := awsclients.LateInitializeFromResponse("", inConfig, respConfig,
+		awsclients.Replacer("ID", "Id"),
+		awsclients.Replacer("ARN", "Arn"),
+		awsclients.MapReplacer(map[string]string{
+			"HTTPVersion": "HttpVersion",
+		}))
+
+	return err
+}
+
+func isUpToDate(cr *svcapitypes.Distribution, gdo *svcsdk.GetDistributionOutput) (bool, error) {
+	patch, err := createPatch(gdo, &cr.Spec.ForProvider)
+
+	if err != nil {
+		return false, err
+	}
+
+	return cmp.Equal(&svcapitypes.DistributionConfig{}, patch, cmpopts.EquateEmpty(),
+		cmpopts.IgnoreTypes(&xpv1.Reference{}, &xpv1.Selector{}, []xpv1.Reference{})), nil
+}
+
+func postObserve(_ context.Context, cr *svcapitypes.Distribution, gdo *svcsdk.GetDistributionOutput,
+	eo managed.ExternalObservation, err error) (managed.ExternalObservation, error) {
+	if err != nil {
+		return managed.ExternalObservation{}, err
+	}
+
+	if awsclients.StringValue(gdo.Distribution.Status) == stateDeployed &&
+		awsclients.BoolValue(gdo.Distribution.DistributionConfig.Enabled) {
+		cr.SetConditions(xpv1.Available())
+	} else {
+		cr.SetConditions(xpv1.Unavailable())
+	}
+
+	return eo, nil
+}
+
+func preUpdate(_ context.Context, cr *svcapitypes.Distribution, udi *svcsdk.UpdateDistributionInput) error {
+	udi.Id = awsclients.String(meta.GetExternalName(cr))
+	udi.SetIfMatch(awsclients.StringValue(cr.Status.AtProvider.ETag))
```

Comment:
Is using `ETag` required and how frequently does it change? If it's as frequent as each request, we might get re-queued constantly.

Reply:
Yes, it's required. If not set, the request is invalid and results in a HTTP-400 error if I don't remember wrong. My observation is that (again if I recall correctly), if the SDK object is updated, it's assigned a new ETag and you need to supply that ETag in your requests.